### PR TITLE
Enable 3d dot

### PR DESCRIPTION
**Pull request by LiyangLingIntel**
This pull request enable 3d dot. 
With this change, we can pass 161 `test_core.py::test_dot3d` unit test cases.

### CODE REVIEW

**Review by etiotto — COMMENTED** [review]
Does the test in test_core.py work now ?

**Comment by LiyangLingIntel** [comment]
> Does the test in test_core.py work now ?

It can take the 3D tensor input and lowering to the end, but the result is incorrect. 
It should be the indices calculation issue, I'm working on figuring them out.

**Comment by LiyangLingIntel** [comment]
This issue is still WIP. Current status for dot3d UT is that it have **85 passes** over **161 cases**.

Most of the remaining failures are caused by these 2 reasons:
* FMA Dot do not support 3d yet. I'm checking why these cases go through FMA Dot, then decide should we apply 3d dot changes to FMA. This would need to change some upstream common file.
* Batch dim offset calculation seems not correct.

**Review by etiotto — COMMENTED** [review]
Some of the skip lists need to be updated (e.g. scripts/skiplist/default/language.txt) like you did for scripts/skiplist/a770/language.txt).

File: third_party/intel/lib/Dialect/TritonIntelGPU/IR/Dialect.cpp [hunk 1]
```diff
@@ -157,65 +186,99 @@ unsigned DpasEncodingAttr::getTotalElemsPerThread(ArrayRef<int64_t> shape,
...
-            std::max<int64_t>(1, shape[1] / shapePerWarp[1])};
+    int64_t numRepBatch =
+        rank == 3 ? std::max<int64_t>(1, shape[0] /
+                                             (shapePerWarp[0] * warpsPerCTA[0]))
+                  : 1;
+    return {numRepBatch,
+            std::max<int64_t>(1, shape[rank - 2] / (shapePerWarp[rank - 2] *
+                                                    warpsPerCTA[rank - 2])),
+            std::max<int64_t>(1, shape[rank - 1] / shapePerWarp[rank - 1])};
   } else if (opIdx == 1) {
```
Comment (etiotto) [inline third_party/intel/lib/Dialect/TritonIntelGPU/IR/Dialect.cpp:223]:
```suggestion
  } 
  
  if (opIdx == 1) {
```

File: third_party/intel/lib/Dialect/TritonIntelGPU/IR/Dialect.cpp [hunk 2]
```diff
@@ -157,65 +186,99 @@ unsigned DpasEncodingAttr::getTotalElemsPerThread(ArrayRef<int64_t> shape,
...
-        std::max<int64_t>(1, shape[1] / (shapePerWarp[1] * warpsPerCTA[1]))};
+    int64_t numRepBatch =
+        rank == 3 ? std::max<int64_t>(1, shape[0] /
+                                             (shapePerWarp[0] * warpsPerCTA[0]))
+                  : 1;
+    return {numRepBatch,
+            std::max<int64_t>(1, shape[rank - 2] / shapePerWarp[rank - 2]),
+            std::max<int64_t>(1, shape[rank - 1] / (shapePerWarp[rank - 1] *
+                                                    warpsPerCTA[rank - 1]))};
   } else {
```
Comment (etiotto) [inline third_party/intel/lib/Dialect/TritonIntelGPU/IR/Dialect.cpp:233]:
```suggestion
  } 
  
```

File: third_party/intel/lib/Dialect/TritonIntelGPU/IR/Dialect.cpp [hunk 3]
```diff
@@ -157,65 +186,99 @@ unsigned DpasEncodingAttr::getTotalElemsPerThread(ArrayRef<int64_t> shape,
...
+        rank == 3 ? std::max<int64_t>(1, shape[0] /
+                                             (shapePerWarp[0] * warpsPerCTA[0]))
+                  : 1;
+    return {numRepBatch,
+            std::max<int64_t>(1, shape[rank - 2] / (shapePerWarp[rank - 2] *
+                                                    warpsPerCTA[rank - 2])),
+            std::max<int64_t>(1, shape[rank - 1] / (shapePerWarp[rank - 1] *
+                                                    warpsPerCTA[rank - 1]))};
   }
+  return rep;
```
Comment (etiotto) [inline third_party/intel/lib/Dialect/TritonIntelGPU/IR/Dialect.cpp:246]:
```suggestion
```

File: third_party/intel/lib/Dialect/TritonIntelGPU/IR/Dialect.cpp [hunk 4]
```diff
@@ -157,65 +186,99 @@ unsigned DpasEncodingAttr::getTotalElemsPerThread(ArrayRef<int64_t> shape,
...
 }
 
 unsigned DpasEncodingAttr::getTotalElemsPerThreadForOperands(
     ArrayRef<int64_t> shape, mlir::Type eltTy, int kWidth, int opIdx) const {
   auto shapePerCTA = getShapePerCTA(*this, shape);
   auto rep = getDPASRepetitions(shapePerCTA, opIdx);
   auto threadsPerWar = getSubGroupSize();
+  int rank = shape.size();
```
Comment (etiotto) [inline third_party/intel/lib/Dialect/TritonIntelGPU/IR/Dialect.cpp:254]:
```suggestion
  size_t rank = shape.size();
```

File: third_party/intel/lib/Dialect/TritonIntelGPU/IR/Dialect.cpp [hunk 5]
```diff
@@ -224,33 +287,47 @@ SmallVector<unsigned> DpasEncodingAttr::getWarpsPerCTA() const {
...
-  auto threadsPerWarp = getThreadsPerWarp();
+  size_t rank = parentShapePerCTATile.size();
   if (opIdx == 0) {
     auto shapeA = getShapeA();
-    return {parentShapePerCTATile[0], shapeA[1]};
+    if (rank == 2)
+      return {parentShapePerCTATile[0], shapeA[1]};
+    else
+      return {parentShapePerCTATile[0], parentShapePerCTATile[rank - 2],
+              shapeA[rank - 1]};
```
Comment (etiotto) [inline third_party/intel/lib/Dialect/TritonIntelGPU/IR/Dialect.cpp:314]:
```suggestion
    return (rank == 2)
       ? {parentShapePerCTATile[0], shapeA[1]}
       : {parentShapePerCTATile[0], parentShapePerCTATile[rank - 2], shapeA[rank - 1]};
```

File: third_party/intel/lib/Dialect/TritonIntelGPU/IR/Dialect.cpp [hunk 6]
```diff
@@ -224,33 +287,47 @@ SmallVector<unsigned> DpasEncodingAttr::getWarpsPerCTA() const {
...
+  res[rank - 1] = executionSize;
+  return res;
 }
 
 SmallVector<unsigned>
 DpasEncodingAttr::getShapePerCTATileForDotOperands(ArrayRef<int64_t> shape,
                                                    int opIdx) const {
   auto parentShapePerCTATile = getShapePerCTATile(shape);
-  auto threadsPerWarp = getThreadsPerWarp();
+  size_t rank = parentShapePerCTATile.size();
```
Comment (etiotto) [inline third_party/intel/lib/Dialect/TritonIntelGPU/IR/Dialect.cpp:307]: Rank here should be 2 or 3 correct? Please add an assertion.

File: third_party/intel/lib/Dialect/TritonIntelGPU/IR/Dialect.cpp [hunk 7]
```diff
@@ -102,51 +102,80 @@ SmallVector<unsigned> DpasEncodingAttr::getDPASInstShapeC() const {
...
 SmallVector<unsigned>
 DpasEncodingAttr::getShapePerCTATile(ArrayRef<int64_t> tensorShape) const {
   auto shapeC = getShapeC();
-  return {shapeC[0] * getWarpsPerCTA()[0], shapeC[1] * getWarpsPerCTA()[1]};
+  SmallVector<unsigned> warpsPerCTA = getWarpsPerCTA();
+  size_t rank = shapeC.size();
+  SmallVector<unsigned> shapePerCTATile(rank);
+  for (size_t i = 0; i < rank; ++i) {
+    shapePerCTATile[i] = shapeC[i] * warpsPerCTA[i];
+  }
```
Comment (victor-eds) [inline third_party/intel/lib/Dialect/TritonIntelGPU/IR/Dialect.cpp:158]:
```suggestion
  llvm::transform(llvm::zip_equal(shapeC, warpsPerCTA), shapePerCTATile.begin(),
       [](auto entry) { return entry.first * entry.second; });
```
Cleaner IMO

File: third_party/intel/lib/Dialect/TritonIntelGPU/IR/Dialect.cpp [hunk 8]
```diff
@@ -157,65 +186,99 @@ unsigned DpasEncodingAttr::getTotalElemsPerThread(ArrayRef<int64_t> shape,
...
+    return (totalElem / threadsPerWar) * product<int64_t>(rep);
   }
+  llvm_unreachable("DpasEncodingAttr opIdx must be 0 or 1");
 }
 
-SmallVector<unsigned> DpasEncodingAttr::getWarpOrder() const { return {1, 0}; }
+SmallVector<unsigned> DpasEncodingAttr::getWarpOrder() const {
+  size_t rank = getWarpsPerCTA().size();
+  SmallVector<unsigned> order(rank);
+  std::iota(order.rbegin(), order.rend(), 0);
```
Comment (victor-eds) [inline third_party/intel/lib/Dialect/TritonIntelGPU/IR/Dialect.cpp:273]: `llvm::to_vector(llvm::reverse(llvm::seq<unsigned>(rank)));` was nicer IMO

File: third_party/intel/lib/Dialect/TritonIntelGPU/IR/Dialect.cpp [hunk 9]
```diff
@@ -157,65 +186,99 @@ unsigned DpasEncodingAttr::getTotalElemsPerThread(ArrayRef<int64_t> shape,
...
+  return order;
+}
 
 SmallVector<unsigned> DpasEncodingAttr::getThreadOrder() const {
-  return {1, 0};
+  size_t rank = getWarpsPerCTA().size();
+  SmallVector<unsigned> order(rank);
+  std::iota(order.rbegin(), order.rend(), 0);
```
Comment (victor-eds) [inline third_party/intel/lib/Dialect/TritonIntelGPU/IR/Dialect.cpp:280]: Same

**Comment by etiotto** [comment]
Latest run shows: 
```
language: passed: 10859, failed: 0, skipped: 170, xfailed: 1017, total: 12046, fixme: 0, pass rate (w/o xfailed): 98.46%
```
The skipped are too many. Can you update the skil lists please ?

**Comment by chengjunlu** [comment]
> Latest run shows:
> 
> ```
> language: passed: 10859, failed: 0, skipped: 170, xfailed: 1017, total: 12046, fixme: 0, pass rate (w/o xfailed): 98.46%
> ```
> 
> The skipped are too many. @LiyangLingIntel can you update all the skip lists please ?

Removed 3d dot cases from the default and xe2 config skip lists.

**Review by victor-eds — APPROVED** [review]

**Comment by etiotto** [comment]
language: passed: 11020, failed: 0, skipped: 9, xfailed: 1017, total: 12046, fixme: 0, pass rate (w/o xfailed): 99.92%

**Review by etiotto — APPROVED** [review]

**Pull request merged**